src/node.zig

```diff
@@ -331,55 +465,147 @@ fn request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_valu
         translate.throw(env, "Unknown operation.") catch return null;
     }
 
+    const message = context.client.get_message() orelse {
+        translate.throw(
+            env,
+            "Message pool has been exhausted.",
+        ) catch return null;
+    };
+    defer context.client.unref(message);
+
     const operation = @intToEnum(Operation, @intCast(u8, operation_int));
-    const events = decode_slice_from_array(env, argv[2], operation) catch |err| switch (err) {
+    const body_length = decode_events(
+        env,
+        argv[2],
+        operation,
+        message.buffer[@sizeOf(Header)..],
+    ) catch |err| switch (err) {
         error.ExceptionThrown => {
             return null;
         },
     };
-    defer allocator.free(events);
 
-    context.client.request(@bitCast(u128, user_data), on_result, operation, events);
+    context.client.request(@bitCast(u128, user_data), on_result, operation, message, body_length);
 
     return null;
 }
 
-fn on_result(user_data: u128, operation: Operation, results: []const u8) void {
+/// The batch has already been encoded into a byte slice. This means that we only have to do one 
+/// copy directly into an available message. No validation of the encoded data is performed except
+/// that it will fit into the message buffer.
+fn raw_request(env: c.napi_env, info: c.napi_callback_info) callconv(.C) c.napi_value {
+    var argc: usize = 20;
+    var argv: [20]c.napi_value = undefined;
+    if (c.napi_get_cb_info(env, info, &argc, &argv, null, null) != .napi_ok) {
+        translate.throw(env, "Failed to get args.") catch return null;
+    }
+
+    if (argc != 4) translate.throw(
+        env,
+        "Function request() requires 4 arguments exactly.",
+    ) catch return null;
+
+    const context_raw = translate.value_external(
+        env,
+        argv[0],
+        "Failed to get Client Context pointer.",
+    ) catch return null;
+    const context = contextCast(context_raw.?) catch return null;
+    const operation_int = translate.u32_from_value(env, argv[1], "operation") catch return null;
+    const user_data = translate.user_data_from_value(env, argv[3]) catch return null;
+
+    if (operation_int >= @typeInfo(Operation).Enum.fields.len) {
+        translate.throw(env, "Unknown operation.") catch return null;
+    }
+    const operation = @intToEnum(Operation, @intCast(u8, operation_int));
+
+    const message = context.client.get_message() orelse {
+        translate.throw(
+            env,
+            "Message pool has been exhausted.",
```

Comment:
```suggestion
            "Too many requests outstanding.",
```

Reply:
I think saying something like this helps the user decide how they can fix this from happening on their end, without needing to know about our message pool internals.